common/model/transfer.c

```diff
@@ -1,676 +1,474 @@
+              __LINE__);
+    return NULL;
+  }
+
+  if (!tag) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] tag cannot be NULL.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  tf = (transfer_t*)calloc(1, sizeof(transfer_t));
+  if (!tf) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  value_out = (transfer_value_out_t*)calloc(1, sizeof(transfer_value_out_t));
+  if (!value_out) {
+    free(tf);
+    return NULL;
+  }
+  memcpy(value_out, output, sizeof(transfer_value_out_t));
+
+  tf->type = VALUE_OUT;
+  tf->address = address;
+  tf->value = value;
+  tf->tag = tag;
+  tf->timestamp = timestamp;
+  tf->meta = (void*)value_out;
+  return tf;
+}
+
+// create a value_int transfer object.
+transfer_t* transfer_value_in_new(flex_trit_t const* const address,
+                                  flex_trit_t const* const tag, int64_t value,
+                                  flex_trit_t const* const data,
+                                  size_t data_len, uint64_t timestamp) {
+  transfer_value_in_t* value_in = NULL;
+  transfer_t* tf = NULL;
+  if (!address) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] address cannot be NULL.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  if (!tag) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] tag cannot be NULL.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  if (!data && data_len != 0) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] data length should be 0.\n",
+              __func__, __LINE__);
+    return NULL;
+  }
+
+  tf = (transfer_t*)calloc(1, sizeof(transfer_t));
+  if (!tf) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  value_in = (transfer_value_in_t*)calloc(1, sizeof(transfer_value_in_t));
+  if (!value_in) {
+    free(tf);
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+  value_in->len = data_len;
+  value_in->data = data;
+
+  tf->type = VALUE_IN;
+  tf->address = address;
+  tf->tag = tag;
+  tf->value = value;
+  tf->timestamp = timestamp;
+  tf->meta = (void*)value_in;
+  return tf;
+}
+
+void transfer_free(transfer_t** transfer) {
+  if (transfer || *transfer) {
+    free((*transfer)->meta);
+    free(*transfer);
+    *transfer = NULL;
+  }
+}
+
+size_t transfer_transactions_count(transfer_t* tf) {
+  transfer_data_t* data = NULL;
+  transfer_value_out_t* value_out = NULL;
+  switch (tf->type) {
+    case DATA:
+      data = (transfer_data_t*)tf->meta;
+      return (data->len + NUM_TRITS_SIGNATURE - 1) / NUM_TRITS_SIGNATURE;
+    case VALUE_OUT:
+      // security level
+      value_out = (transfer_value_out_t*)tf->meta;
+      return value_out->security;
+    case VALUE_IN:
+      // fixed
+      return 1;
+  }
+  return 0;
+}
+
+transfer_ctx_t* transfer_ctx_new() {
```

Comment:
Do we need dynamic allocation for the ctx ?

Reply:
not sure, but just keep the logic in the previous change.

Reply:
I think we don't need it

Reply:
btw, why it exists?

Reply:
I have no idea :)

Reply:
then, why do you think we don't need it?

Reply:
Looking where it's being use (e.g. `transfer_iterator_new`), I don't understand why it's being dynamically allocated

Reply:
I prefer not to change something doesn't fully understand, we can seek for the answer from @th0br0 and @ifullgaz.

Reply:
Feel free to remove the method, it was an experiment. But in hindsight, I'm not that happy with the pointer aliases either.